server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -298,22 +305,173 @@ private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecently(Master
+                );
+            }
+        }
+        Map<DiscoveryNode, ClusterFormationFailureHelper.ClusterFormationState> nodeClusterFormationStateMap =
+            nodeToClusterFormationStateOrExceptionMapCopy.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().clusterFormationState()));
+        if (anyNodeInClusterReportsDiscoveryProblems(masterEligibleNodes, nodeClusterFormationStateMap)) {
+            result = new CoordinationDiagnosticsResult(
+                CoordinationDiagnosticsStatus.RED,
+                String.format(
+                    Locale.ROOT,
+                    "No master node observed in the last %s, and some master eligible nodes are unable to discover other master "
+                        + "eligible nodes",
+                    nodeHasMasterLookupTimeframe
+                ),
+                getDetails(explain, localMasterHistory, null, coordinator.getClusterFormationState().getDescription())
+            );
+        } else {
+            if (anyNodeInClusterReportsQuorumProblems(nodeClusterFormationStateMap)) {
+                result = new CoordinationDiagnosticsResult(
+                    CoordinationDiagnosticsStatus.RED,
+                    String.format(
+                        Locale.ROOT,
+                        "No master node observed in the last %s, and the master eligible nodes are unable to form a quorum",
+                        nodeHasMasterLookupTimeframe
+                    ),
+                    getDetails(explain, localMasterHistory, null, coordinator.getClusterFormationState().getDescription())
+                );
+            } else {
+                result = new CoordinationDiagnosticsResult(
+                    CoordinationDiagnosticsStatus.RED,
+                    String.format(
+                        Locale.ROOT,
+                        "No master node observed in the last %s, and the cause has not been determined.",
+                        nodeHasMasterLookupTimeframe
+                    ),
+                    getDetails(explain, localMasterHistory, null, coordinator.getClusterFormationState().getDescription())
+                );
+            }
         }
         return result;
     }
 
+    /**
+     * This method checks whether each master eligible node has discovered each of the other master eligible nodes. If any node in
+     * nodeToClusterFormationStateMap to discover any other node in masterEligibleNodes, then this method returns true. If no discovery
+     * problems are detected, this method returns false.
+     * @param masterEligibleNodes The collection of all master eligible nodes
+     * @param nodeToClusterFormationStateMap A map of each master node to its ClusterFormationState
+     * @return true if there are discovery problems, false otherwise
+     */
+    private boolean anyNodeInClusterReportsDiscoveryProblems(
```

Comment:
I think this is a bit confusing and would benefit from possibly being static (if possible), package protected and unit tested.
`discoveryProblems` is very generic IMO - what problems is it looking for? Cannot connect to each other ?

Reply:
This one is currently unit tested by `testRedForDiscoveryProblems`. Do you think there's a need to crack into the private method?

Reply:
Any thoughts on what to call this? It's looking for any node in the cluster whose `coordinator.getClusterFormationState().foundPeers()` does not contain everything in my local `coordinator.getFoundPeers()` (plus myself if I'm master-eligible). So it's probably a join problem, but we don't know for sure. It's hard to get all of that into a method name that's not a paragraph. :)